### PR TITLE
Log compaction

### PR DESCRIPTION
**Pull request by liw**

### CODE REVIEW

**Comment by liw** [comment]
@okulkarni, although the integration testing is not yet done, would you mind starting the reviews anyway?

File: src/raft_server.c [hunk 1]
```diff
@@ -1293,26 +1293,19 @@ int raft_begin_load_snapshot(
     if (last_included_index == 0 || last_included_index == 0)
```
Comment (okulkarni) [inline src/raft_server.c]: Not part of your change, but the two conditions are identical.

File: README.rst [hunk 2]
```diff
@@ -432,20 +432,22 @@ Log Compaction
...
-10. Finally, peer calls ``raft_node_set_active`` to nodes as per the snapshot's membership info.
+5. When the ``send_installsnapshot`` callback fires, the user must propogate a chunk of the snapshot to the other node by ``msg_installsnapshot_t`` and implementation-specific additional arguments (e.g., the offset of the chunk).
+6. When the peer receives the chunk, the user must call ``raft_recv_installsnapshot``. When ``recv_installsnapshot`` fires, the user must process the chunk and fill any implementation-specific arguments to the response.
+7. When the ``recv_installsnapshot_response`` callback fires, the user must record the progress of the snapshot transfer, typically in the user data of the ``raft_node_t`` object for the peer.
+8. Once the peer has the complete snapshot, the user must call ``raft_begin_load_snapshot``.
+9. Peer calls ``raft_add_node`` to add nodes as per the snapshot's membership info.
+10. Peer calls ``raft_node_set_voting`` to nodes as per the snapshot's membership info.
+11. Peer calls ``raft_node_set_active`` to nodes as per the snapshot's membership info.
+12. Finally, peer calls ``raft_node_set_active`` to nodes as per the snapshot's membership info.
```
Comment (okulkarni) [inline README.rst]: Are 11 and 12 saying the exact same thing?

File: include/raft.h [hunk 3]
```diff
@@ -223,20 +249,58 @@ typedef int (
...
     void *user_data,
-    raft_node_t* node
+    raft_node_t* node,
+    msg_installsnapshot_t* msg
+    );
+
+/** Callback for receiving InstallSnapshot request messages.
+ * @param[in] raft The Raft server making this callback
+ * @param[in] user_data User data that is passed from Raft server
+ * @param[in] node The node's ID that we are sending this message to
```
Comment (okulkarni) [inline include/raft.h]: This should be the node we are receiving from.

File: README.rst (hunk 2 above)
Comment (liw) [reply README.rst]: Oops. Will fix.

File: src/raft_server.c [hunk 4]
```diff
@@ -853,6 +952,18 @@ raft_entry_t* raft_get_entries_from_idx(raft_server_t* me_, int idx, int* n_etys
...
+int raft_send_installsnapshot(raft_server_t* me_, raft_node_t* node)
+{
+    raft_server_private_t* me = (raft_server_private_t*)me_;
+
+    msg_installsnapshot_t is = {};
+    is.term = me->current_term;
+    is.last_idx = log_get_base(me->log);
+    is.last_term = log_get_base_term(me->log);
+
+    return me->cb.send_installsnapshot(me_, me->udata, node, &is);
```
[the remark this answers is not in the document]
Comment (liw) [reply src/raft_server.c]: Perhaps we could assemble these checks into raft_set_callbacks in the future.

File: include/raft.h (hunk 3 above)
Comment (liw) [reply include/raft.h]: The node object, of course. I'll fixed this for the new callbacks as well as the existing ones.

File: src/raft_server.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (okulkarni) [reply src/raft_server.c]: That's fine.. the only problem is if any of these methods get called before set_callbacks, we may run into a problem. So probably still need to add asserts.

[the remark this answers is not in the document]
Comment (liw) [reply src/raft_server.c]: Okay... I'll assert each before calling.

**Comment by liw** [comment]
Refreshed "log: Remove RAFT_LOGTYPE_SNAPSHOT" to fix raft_get_last_log_term to rely on raft_get_entry_term. It used to directly get the current index, which may be the base.

Added "server: Fix node pointer-ID confusion" and "server: Don't remove nodes when loading snapshots". The first one addresses Omkar's previous comments on API documentation. The second one addresses the raft_node_t and user data memory leak.

**Pull request merged**